### ISSUE TITLE
6. Regression_sklearn_LogisticRegression

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_python-II#30]
"""
sklearn 로지스틱 회귀모델
 - y변수가 범주형인 경우
"""

from sklearn.datasets import load_breast_cancer, load_iris # dataset
from sklearn.linear_model import LogisticRegression # model
from sklearn.metrics import accuracy_score, confusion_matrix # model 평가

### 1. dataset load & 변수선택
breast = load_breast_cancer()

X= breast.data # x 변수 # 대문자는 차원이 커서 그냥 구별할려고
y = breast.target # y 변수

X.shape # (569, 30)
y.shape # (569,)

### 2. model 생성
help(LogisticRegression)
'''
random_state=None : 난수 seed 값 지정
solver='lbfgs' : 알고리즘 
- solver : {'newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'},default='lbfgs'
max_iter =100 : 반복학습 횟수
multi_class='auto' : 다항분류
- multi_class : {'auto', 'ovr', 'multinomial'}, default='auto'

적용 예)
일반 데이터, 이항분류 : default
일반 데이터, 다항분류 : solver='lbfgs', multi_class=multinomial
빅데이터, 이항분류 : solver='sag' or'saga'
빅데이터, 다항분류 : solver='sag' or'saga', multi_class=multinomial
'''
lr = LogisticRegression(random_state=123)
model = lr.fit(X=X, y=y)
model # multi_class='auto' -> sigmoid 활성함수이용 -> 이항분류
'''
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=123, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)
'''

### 3. model 평가
-acc 방법1)
acc = model.score(X, y)
print('accuracy =', acc) # accuracy = 0.9472759226713533 상당히 높다. 94%
-acc 방법2)
y_pred = model.predict(X)
acc = accuracy_score(y,y_pred)
acc # 0.9472759226713533

con_max = confusion_matrix(y,y_pred)
type(con_max) # numpy.ndarray에서 구성요소 꺼내는 법 con_max[0,0]
print(con_max)
'''
     0    1
0 [[193  19]
1 [ 11 346]]
'''
acc = (con_max[0,0] + con_max[1,1])/con_max.sum()
acc # 0.9472759226713533

tab = pd.crosstab(y, y_pred, rownames=['관측치'],colnames = ['예측치'])
'''
예측치    0    1
관측치          
0    193   19
1     11  346
'''
-판다스 객체라서 loc
acc = (tab.loc[0,0] + tab.loc[1,1])/ len(y)
acc  # 0.9472759226713533

### 1. dataset load
iris= load_iris()
iris.target_names # array(['setosa', 'versicolor', 'virginica'], dtype='<U10')
X,y = load_iris(return_X_y = True)
 
X.shape # (150, 4)
y.shape # (150,)
y # 0 ~ 2

### 2. model 생성
lr = LogisticRegression(random_state=123,solver='lbfgs', multi_class='multinomial')
- 'multinomial' : softmax 활성함수 사용 -> 다항분류

'''
sigmoid function : 0~1 확률값 -> cutoff = 0.5 -> 이상이면 1로, 아니면 0으로
softmax function :  0~1 확률값 -> 전체합 = 1(c0:0.1, c1 : 0.1,c2:0.8) -> 제일 큰 c3으로다항분류
'''
model = lr.fit(X,y)
y_pred = model.predict(X) # class로 예측
'''
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])
'''
y_pred2 = model.predict_proba(X) # 확률값
'''
array([[9.81797141e-01, 1.82028445e-02, 1.44269293e-08],
       [9.71725476e-01, 2.82744937e-02, 3.01659208e-08],
       [9.85444223e-01, 1.45557643e-02, 1.23263078e-08],
       [9.76282998e-01, 2.37169623e-02, 3.97229604e-08],
'''

-['setosa','versicolor','virginica'] = 1
import numpy as np
arr=np.array([9.81797141e-01, 1.82028445e-02, 1.44269293e-08])
arr.max() # 0.981797141 -> setosa 로!
arr.min()
arr.sum() #  0.9999999999269293 = 1

### 3. model 평가
acc = accuracy_score(y,y_pred)
acc # 0.9733333333333334

con_mat = confusion_matrix(y,y_pred)
con_mat
'''
array([[50,  0,  0],
       [ 0, 47,  3],
       [ 0,  1, 49]], dtype=int64)
'''

acc = (con_mat[0,0] + con_mat[1,1] + con_mat[2,2])/con_mat.sum()
acc

### 히트맵 : 시각화
import seaborn as sn # heatmap - Accuracy Score

- confusion matrix heatmap 
plt.figure(figsize=(6,6)) # chart size
sn.heatmap(con_mat, annot=True, fmt=".3f", linewidths=.5, square = True);# , cmap = 'Blues_r' : map »ö»ó 
plt.ylabel('Actual label');
plt.xlabel('Predicted label');
all_sample_title = 'Accuracy Score: {0}'.format(acc)
plt.title(all_sample_title, size = 18)
plt.show()

### 3. digits :  multi class

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

### 1. dataset load
digits = load_digits()
digits.target_names # array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

X = digits.data
y = digits.target
X.shape # (1797, 64) -> 1797장의 images
y.shape # (1797,) ->  1797장의 10진수 정답

### 2. split
img_train, img_test, label_train, label_test = train_test_split(X,y, test_size = 0.25)

-훈련셋 image -> reshape
img2d = img_train.reshape(-1,8,8) # (-1 : 전체 image, )
img2d.shape # (1347, 8, 8)
plt.imshow(img2d[0])
label_train[0] # 2
img2d[0]
'''

-바깥검은색이(0)이 하얀색
-색깔있는 것(13,16,...)이 검은색 # 엑스레이의 원리
array([[ 0.,  3., 13., 16.,  9.,  0.,  0.,  0.],
       [ 0., 10., 15., 13., 15.,  2.,  0.,  0.],
       [ 0., 15.,  4.,  4., 16.,  1.,  0.,  0.],
       [ 0.,  0.,  0.,  5., 16.,  2.,  0.,  0.],
       [ 0.,  0.,  1., 14., 13.,  0.,  0.,  0.],
       [ 0.,  0., 10., 16.,  5.,  0.,  0.,  0.],
       [ 0.,  4., 16., 13.,  8., 10.,  9.,  1.],
       [ 0.,  2., 16., 16., 14., 12.,  9.,  1.]])
'''
img_test.shape # (450, 64)

### 2. model 생성
-모델을 생성할 때는 reshape 없이 그냥 원래 데이터를 넣으면 된다.
lr = LogisticRegression(random_state=123,solver='lbfgs',
                        multi_class='multinomial')
model = lr.fit(img_train,label_train)

y_pred = model.predict(img_test)

### 3. model 평가
acc = accuracy_score(label_test,y_pred)
acc #  0.9511111111111111

con_mat = confusion_matrix(label_test,y_pred)
con_mat
'''
array([[37,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 40,  0,  0,  0,  0,  0,  0,  2,  1],
       [ 0,  1, 40,  3,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 43,  0,  0,  0,  0,  1,  1],
       [ 0,  0,  0,  0, 37,  0,  0,  1,  0,  0],
       [ 0,  0,  0,  0,  0, 46,  0,  0,  0,  2],
       [ 0,  1,  0,  0,  0,  0, 51,  0,  0,  0],
       [ 0,  0,  0,  1,  1,  0,  0, 46,  0,  0],
       [ 0,  3,  1,  0,  0,  0,  0,  0, 43,  1],
       [ 0,  0,  0,  0,  0,  1,  0,  0,  1, 45]], dtype=int64)
'''

result = label_test ==y_pred
result
len(result) # 450

img_test[result, :]

- 틀린 image 살펴보기
false_img = img_test[result==False]
false_img.shape # (22, 64)

for img in false_img : # row
    img2d = img.reshape(8,8)
    plt.imshow(img2d)
    
- 히트맵 : 시각화
import seaborn as sn # heatmap - Accuracy Score

- confusion matrix heatmap 
plt.figure(figsize=(6,6)) # chart size
sn.heatmap(con_mat, annot=True, fmt=".3f", linewidths=.5, square = True);# , cmap = 'Blues_r' : map »ö»ó 
plt.ylabel('Actual label');
plt.xlabel('Predicted label');
all_sample_title = 'Accuracy Score: {0}'.format(acc)
plt.title(all_sample_title, size = 18)
plt.show()
